packages/next/src/server/app-render/app-render.tsx

```diff
@@ -239,11 +240,19 @@ function makeGetDynamicParamFromSegment(
         segmentParam.type === 'optional-catchall' ||
         segmentParam.type === 'catchall'
       ) {
-        // If we weren't able to match the segment to a URL param, and we have a catch-all route,
-        // provide all of the known params (in array format) to the route
-        // It should be safe to assume the order of these params is consistent with the order of the segments.
-        // However, if not, we could re-parse the `pagePath` with `getRouteRegex` and iterate over the positional order.
-        value = Object.values(params).map((i) => encodeURIComponent(i))
+        // derive the value from the page path, replacing any dynamic params with the actual values
+        value = pagePath
+          .split('/')
+          .slice(1)
+          .map((part) => {
+            const match = /^\[(.+)\]$/.exec(part)
+
+            if (match) {
+              return params[match[1]]
+            }
+
+            return part
+          })
```

Comment:
If there's a known utility, please point me to it and I'll update the code here

Reply:
We discussed offline, but [`getRouteRegex`](https://github.com/vercel/next.js/blob/d62b68fb041da77509a82e62afd3e275c060562f/packages/next/src/shared/lib/router/utils/route-regex.ts#L74) would be a good one to use here as it'll also help you determine if it was an optional param or not.